### PR TITLE
[TwigBundle] Add default templates directory and option to configure it

### PR DESCRIPTION
**Pull request by yceruto**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Feature freeze is coming so this one should be important for the new structure. Moving forward and alternative of https://github.com/symfony/symfony/pull/23339 but I'm proposing `templates/bundles/<BundleName>` instead of `templates/bundles/<BundleTwigNamespace>` to override bundles templates and easy migration from current `app/Resources/<BundleName>/views` convention. Also this fix the pending comments.

Summary:
 * Added new option to configure default path for templates directory:
```yaml
twig:
    default_path: '%kernel.project_dir%/templates' # default
```
 * Added new path convention to override bundle templates  `<default_path>/bundles/<BundleName>`:
```
# Examples:
templates/bundles/TwigBundle/Exception/error.html.twig - @Twig/Exception/error.html.twig
templates/bundles/FOSUserBundle/layout.html.twig - @FOSUser/layout.html.twig
```

Current templates in `<kernel.root_dir>/Resources/<BundleName>/views` have priority over the new one, and both have priority over the bundle `views` path.

### CODE REVIEW

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php [hunk 1]
```diff
@@ -192,6 +197,11 @@ private function getBundleHierarchy(ContainerBuilder $container)
             }
             $container->addResource(new FileExistenceResource($dir));
 
+            if (file_exists($dir = $container->getParameterBag()->resolveValue($config['default_path']).'/overrides/'.$name)) {
```
Comment (ro0NL) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/TwigExtension.php]: cant we just move the sub path (`/overrides/`) to the config layer? Id say `%kernel.project_dir%/templates/bundles` is a sane default.

Not sure why you prefer "overrides" (to vague to me).

**Comment by mvrhov** [comment]
Please no overrides. just keep it inside a bundles.

**Comment by yceruto** [comment]
Changed to `bundles` directory.

File: src/Symfony/Bundle/TwigBundle/DependencyInjection/Configuration.php [hunk 2]
```diff
@@ -158,6 +158,9 @@ private function addTwigOptions(ArrayNodeDefinition $rootNode)
                     ->end()
                     ->prototype('variable')->end()
                 ->end()
+                ->scalarNode('default_path')
+                    ->defaultValue('%kernel.project_dir%/templates')
+                ->end()
```
Comment (chalasr) [inline src/Symfony/Bundle/TwigBundle/DependencyInjection/Configuration.php]: I would move it just before `paths`, maybe document it using `info()` also

**Review by chalasr — COMMENTED** [review]
We should have a test case for this.

**Comment by yceruto** [comment]
Tests added.

**Comment by ro0NL** [comment]
Just a question; should we deprecate the old paths (hardcoded in src).

Ie trigger in case of `if (file_exists($dir = $container->getParameter('kernel.root_dir').'/Resources/views')) {` etc. Wouldnt that me more concise?

**Review by fabpot — APPROVED** [review]

**Review by chalasr — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @yceruto.

**Comment by fabpot** [comment]
Just tested on 2 of my projects, and it works perfectly.

**Comment by Tobion** [comment]
I agree with @ro0NL that we should deprecate the old `/Resources/<BundleName>/views`.

1. it's not good to have different ways for the same thing as makes things inconsistent
2. there should be a configuration for the new paths instead of hardcoding it as it makes things self-explaining if you add that to the default config file. for the same reason we removed auto-registration of commands in favor of having an explicit import (psr-4 loading and autoconfiguration or not).

**Comment by ro0NL** [comment]
@Tobion at time of posting i didnt realize we need to support both variants; legacy and new projects.

https://github.com/symfony/symfony/issues/24227#issuecomment-329942775

I think this is also related we dont want to force users switching to a new structure for no real benefit. But the fact you can mix&match can be confusing for sure! So yes, i think the current approach is a bit too pragmatic.

**Comment by nesl247** [comment]
Can anyone tell me if this was tested / supposed to work with Form templates? Doesn't seem to be.

**Comment by xabbuh** [comment]
@nesl247 please open a new issue if you think you discovered a bug

**Comment by nesl247** [comment]
Will do. Wasn't sure if I should or not as I'm not sure if it's working as intended or a bug.

**Comment by yceruto** [comment]
> When I have an old reference to the template, with colon notation, it's not possible to override template in a /templates folder by design. Should override be done in src/Resources/.. folder only?
So to use /templates folder to override, the only option I have - rename all the template references?

That's correct. 

> Asking, because we have more than 2k templates references, and renaming could cause new bugs, but maintaining two folders for override templates at the same time is confusing for developers.

Since 4.2 `src/Resources/views` is deprecated but still works as Twig's path and has priority over `templates/`dir, so first, you can change the notation of the template name without breaking the loading process (thus, changing of loader and making use Twig features), and then, move them to `templates/` shouldn't break anything either (it's just an exchange of paths).

**Pull request merged**